tests/test_sftp.py

```diff
@@ -0,0 +1,341 @@
+
+    def close(self):
+        self.closed = True
+
+
+# ============== Tests for SFTPDelivery ==============
+
+
+def test_sftp_delivery_valid_initialization(sftp_delivery):
+    assert sftp_delivery.host == "sftp.example.com"
+    assert sftp_delivery.port == 22
+    assert sftp_delivery.username == "user"
+    assert sftp_delivery.password == "password"
+    assert sftp_delivery.private_key is None
+
+
+def test_sftp_delivery_invalid_host():
+    with pytest.raises(TypeError):
+        SFTPDelivery(host=123, username="user", password="password")
+    with pytest.raises(ValueError):
+        SFTPDelivery(host="   ", username="user", password="password")
+
+
+def test_sftp_delivery_invalid_port():
+    with pytest.raises(TypeError):
+        SFTPDelivery(
+            host="sftp.example.com", username="user", password="password", port="22"
+        )
+    with pytest.raises(ValueError):
+        SFTPDelivery(
+            host="sftp.example.com", username="user", password="password", port=70000
+        )
+
+
+def test_sftp_delivery_invalid_username():
+    with pytest.raises(TypeError):
+        SFTPDelivery(host="sftp.example.com", username=123, password="password")
+    with pytest.raises(ValueError):
+        SFTPDelivery(host="sftp.example.com", username="", password="password")
+
+
+def test_sftp_delivery_invalid_password():
+    with pytest.raises(TypeError):
+        SFTPDelivery(host="sftp.example.com", username="user", password=123)
+    with pytest.raises(ValueError):
+        SFTPDelivery(host="sftp.example.com", username="user", password="")
+
+
+def test_sftp_delivery_invalid_private_key():
+    with pytest.raises(TypeError):
+        SFTPDelivery(
+            host="sftp.example.com",
+            username="user",
+            password="password",
+            private_key=123,
+        )
+    with pytest.raises(ValueError):
+        SFTPDelivery(
+            host="sftp.example.com",
+            username="user",
+            password="password",
+            private_key="   ",
+        )
+
+
+def test_sftp_delivery_connect(monkeypatch, sftp_delivery):
+    from quipus.services import sftp_delivery as sftp_module
+
+    monkeypatch.setattr("paramiko.SSHClient", MockSSHClient)
+    monkeypatch.setattr(sftp_module, "SFTPClient", MockSFTPClient)
+
+    sftp_delivery.connect()
+    assert sftp_delivery.connection.connected is True
+    assert sftp_delivery.sftp_client is not None
+
+
+def test_sftp_delivery_connect_invalid_credentials(monkeypatch, sftp_delivery):
+    class MockSSHClientFailure(MockSSHClient):
+        def connect(self, hostname, port, username, password=None, pkey=None):
+            raise paramiko.AuthenticationException("Authentication failed.")
+
+    monkeypatch.setattr("paramiko.SSHClient", MockSSHClientFailure)
+
+    with pytest.raises(paramiko.AuthenticationException):
+        sftp_delivery.connect()
+
+
+def test_sftp_delivery_upload_file(monkeypatch, sftp_delivery, tmp_path):
```

Comment:
This test should check that `put` has been called with `local_file` and `remote_file` args.